src/ProductManifold.jl

```diff
@@ -0,0 +1,476 @@
+expects two-by-two matrices, so its size specification is `Tuple{2,2}`.
+"""
+struct ShapeSpecification{TRanges, TSizes} end
+
+function ShapeSpecification(manifolds::Manifold...)
+    sizes = map(m -> representation_size(m, MPoint), manifolds)
+    lengths = map(prod, sizes)
+    ranges = UnitRange{Int64}[]
+    k = 1
+    for len ∈ lengths
+        push!(ranges, k:(k+len-1))
+        k += len
+    end
+    TRanges = tuple(ranges...)
+    TSizes = Tuple{map(s -> Tuple{s...}, sizes)...}
+    return ShapeSpecification{TRanges, TSizes}()
+end
+
+"""
+    ProductArray(shape::ShapeSpecification, data)
+
+An array-based representation for points and tangent vectors on the
+product manifold. `data` contains underlying representation of points
+arranged according to `TRanges` and `TSizes` from `shape`.
+Internal views for each specific sub-point are created and stored in `parts`.
+"""
+struct ProductArray{TM<:ShapeSpecification,T,N,TData<:AbstractArray{T,N},TV<:Tuple} <: AbstractArray{T,N}
+    data::TData
+    parts::TV
+end
+
+# The two-argument version of this constructor is substantially faster than
+# the generic one.
+function ProductArray(M::Type{ShapeSpecification{TRanges, Tuple{Size1, Size2}}}, data::TData) where {TRanges, Size1, Size2, T, N, TData<:AbstractArray{T,N}}
+    views = (SizedAbstractArray{Size1}(view(data, TRanges[1])),
+             SizedAbstractArray{Size2}(view(data, TRanges[2])))
+    return ProductArray{M, T, N, TData, typeof(views)}(data, views)
+end
+
+function ProductArray(M::Type{ShapeSpecification{TRanges, Tuple{Size1, Size2, Size3}}}, data::TData) where {TRanges, Size1, Size2, Size3, T, N, TData<:AbstractArray{T,N}}
+    views = (SizedAbstractArray{Size1}(view(data, TRanges[1])),
+             SizedAbstractArray{Size2}(view(data, TRanges[2])),
+             SizedAbstractArray{Size3}(view(data, TRanges[3])))
+    return ProductArray{M, T, N, TData, typeof(views)}(data, views)
+end
+
+function ProductArray(M::ShapeSpecification{TRanges, TSizes}, data::TData) where {TM, TRanges, TSizes, T, N, TData<:AbstractArray{T,N}}
+    return ProductArray(typeof(M), data)
+end
+
+@doc doc"""
+    prod_point(M::ShapeSpecification, pts...)
+
+Construct a product point from product manifold `M` based on point `pts`
+represented by [`ProductArray`](@ref).
+
+# Example
+To construct a point on the product manifold $S^2 \times \mathbb{R}^2$
+from points on the sphere and in the euclidean space represented by,
+respectively, `[1.0, 0.0, 0.0]` and `[-3.0, 2.0]` you need to construct shape
+specification first. It describes how linear storage of `ProductArray`
+corresponds to array representations expected by `Sphere(2)` and `Euclidean(2)`.
+
+    M1 = Sphere(2)
+    M2 = Euclidean(2)
+    Mshape = Manifolds.ShapeSpecification(M1, M2)
+
+Next, the desired point on the product manifold can be obtained by calling
+`Manifolds.prod_point(Mshape, [1.0, 0.0, 0.0], [-3.0, 2.0])`.
+"""
+function prod_point(M::ShapeSpecification, pts...)
```

Comment:
Would be nice to bring this inline with `submanifold_component` or whatever we decide, since this isn't unique to points.